System indices treated as restricted indices

jaymode opened the pull request:
System indices should be treated as a special set of indices and not be
accessible by all users. The existing security codebase has the notion
of restricted indices, which are currently a subset of system indices.

This change unifies the two concepts by making system indices the set
of restricted indices. This means that going forward, consumers of
system indices will need access to restricted indices.

Closes #69298

jaymode committed:
System indices treated as restricted indices

System indices should be treated as a special set of indices and not be
accessible by all users. The existing security codebase has the notion
of restricted indices, which are currently a subset of system indices.

This change unifies the two concepts by making system indices the set
of restricted indices. This means that going forward, consumers of
system indices will need access to restricted indices.

Closes #69298

jaymode committed:
restricted index updates

jaymode committed:
Merge remote-tracking branch 'origin/master' into system_indices_restricted_indices

jaymode committed:
admin role needs restricted index access for now

jaymode committed:
fix test failures

jaymode committed:
fix test

jaymode committed:
Revert "fix test"

This reverts commit b12116bf8464899bfd960eae96f0b63ebfe02b87.

jaymode committed:
remove some of restrictedindicesnames

jaymode committed:
fixes

jaymode committed:
fix bad version name

jaymode committed:
unused

jaymode committed:
fix test cp

jaymode committed:
Merge branch 'master' into system_indices_restricted_indices

jaymode commented:
@elasticmachine run elasticsearch-ci/part-1

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

williamrandolph commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:
```diff
@@ -1348,6 +1367,89 @@ private Authentication createAuthentication() {
                 AuthenticationType.ANONYMOUS), Collections.emptyMap());
     }
 
+    public void testXPackUserCanAccessNonSecurityIndices() {
```
Would `testXPackUserCanAccessNonRestrictedIndices` be more accurate?

williamrandolph commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:
```diff
@@ -1348,6 +1367,89 @@ private Authentication createAuthentication() {
+                Matchers.is(false));
+        }
+    }
+
```
test cases to add: async search can't read/write audit trail, does xpack user have cluster privileges?

williamrandolph reviewed (changes requested):
This makes a lot of sense to me.

Much of the code has the goal of replacing a hard-coded list of restricted indices with one that's constructed at node startup. IndexNameExpressionResolver supplies the Security plugin with an automaton that matches all system names, and this automaton is passed to the CompositeRolesStore. The CompositeRolesStore, in turn, constructs every IndicesPermission.Group with a reference to this automaton.

When the application is running, the check for restricted indices happens in the IndicesPermission#checkResourcePrivileges method. (I'm glad that there seems to be one place that handles all of the privilege checking: less potential for bugs.)

Test coverage is good, although I don't see that we have extended test coverage beyond the previous group of restricted indices (security- and async-search-related). I'd like to look at whether we can test that other system indices are treated as restricted.

Since I'm taking this work over from Jay, I will work with Gordon to get this PR merged.

rdnm commented:
@tvernum and @albertzaharovits - @williamrandolph is going to address some comments and get @gwbrown to review them. However, a broader security review would be appreciated.

gwbrown commented on client/rest-high-level/roles.yml:
```diff
@@ -3,6 +3,7 @@ admin:
     - all
   indices:
     - names: '*'
+      allow_restricted_indices: true
```
++

gwbrown commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -317,6 +330,19 @@ private static Automaton featureToIndexAutomaton(Feature feature) {
         return systemIndexAutomaton.orElse(EMPTY);
     }
 
+    private static Automaton buildDataStreamAutomaton(Map<String, Feature> descriptors) {
```
I agree, I'd prefer to not have two different implementations like this. If having them separate is necessary we should clarify why and how they're different.

The only obvious difference to me is that the `Predicate` version does not support patterns - i.e., that predicate will only return `true` if there's an exact match of the data stream name. I'm not sure what, if anything, is using a pattern for the data stream name.

williamrandolph replied:
Since both implementations were built from the same Java stream of system data stream names, I've reimplemented the predicate on top of the automaton.

AthenaEryma reviewed (commented):
Left one minor comment - I didn't otherwise see anything beyond what William's already called out.

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
Improve automaton build stream

williamrandolph commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -317,6 +330,19 @@ private static Automaton featureToIndexAutomaton(Feature feature) {
+                    .map(SystemDataStreamDescriptor::getDataStreamName)
+                    .map(dsName -> SystemIndexDescriptor.buildAutomaton(dsName, null))
+                    .reduce(Operations::union)
+                    .orElse(EMPTY))
```
Good call! Done.

williamrandolph committed:
Handle warnings in tests

williamrandolph commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java:
```diff
@@ -23,21 +20,10 @@
-    public static final Automaton NAMES_AUTOMATON = Automatons.unionAndMinimize(Arrays.asList(Automatons.patterns(RESTRICTED_NAMES),
-            ASYNC_SEARCH_AUTOMATON));
-
     private RestrictedIndicesNames() {
```
Looking at this again, I think it should be a follow-up. This class is used in a lot of tests and refactoring now would clutter up an already-large PR.

williamrandolph committed:
Add some javadoc

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
DRY by using automaton in constructing a predicate

williamrandolph committed:
Refactor to avoid repeating code

williamrandolph committed:
Rename test for accuracy

albertzaharovits commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -763,6 +764,10 @@ boolean isPatternMatchingAllIndices(Metadata metadata, String[] indicesOrAliases
         return false;
     }
 
+    public boolean isSystemName(String name) {
```
This is not used anywhere.

williamrandolph replied:
Removed!

albertzaharovits committed:
Give IndicesPermissions an automaton reference

It's not great that the IndicesPermission has to refer to one of the
restrictedNamesAutomaton from one of its groups.

Instead, we assign the role's restrictedNamesAutomaton to the individual
groups as well as to the containing IndicesPermission.

Author:    Albert Zaharovits <albert.zaharovits@elastic.co>
Date:      Sun Aug 1 13:14:28 2021 +0300

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:
```diff
@@ -297,7 +342,9 @@ private boolean isConcreteRestrictedIndex(String indexPattern) {
         }
-        return RestrictedIndicesNames.isRestricted(indexPattern);
+        CharacterRunAutomaton runAutomaton =
+            new CharacterRunAutomaton(Arrays.stream(groups).map(g -> g.restrictedNamesAutomaton).findFirst().orElse(Automatons.EMPTY));
```
I think it's not great that the `IndicesPermission` has to refer to one of the `restrictedNamesAutomaton` from one of its groups (This is not the only place where this happens place.)

I have tried my hand at avoiding this: https://github.com/jaymode/elasticsearch/compare/system_indices_restricted_indices...albertzaharovits:indices_permission_and_group_share_restricted_indices?expand=1

It works by actually assigning the role's `restrictedNamesAutomaton` to both the individual groups as well as to the containing `IndicesPermission`.

albertzaharovits replied:
Also, `isConcreteRestrictedIndex` is going to create a `CharacterRunAutomaton` on every invocation, which I think can be easily avoided by precomputing.

williamrandolph replied:
I brought in your commit and updated tests to match.

albertzaharovits reviewed (commented):
This looks like a quite safe refactoring overall!

I only have [one suggestion](https://github.com/elastic/elasticsearch/pull/74212/files?file-filters%5B%5D=.gradle&file-filters%5B%5D=.java#r680489122) that we should pass the restricted indices automaton to the `IndicesPermission` since, unfortunately, it needs it as well (in addition to the groups that it contains), and iterating the groups to take the first one or ensuring that all the groups use the same automaton is ugly; it is better to pass the automaton to the `IndicesPermission` and let it pass it along to the groups that it contains.
Let me know if this sounds reasonable to you. I'll take a quick look afterwards and approve.

Not to be covered by this PR, but do we plan to prevent users from creating restricted indices by mistake?

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph commented:
@albertzaharovits

> do we plan to prevent users from creating restricted indices by mistake?

The only way for a user to create a system/restricted index is to define it in a custom plugin, so I don't think we need to worry about accidental restricted indices.

williamrandolph committed:
Remove unused method

williamrandolph committed:
Adjust tests for new builder

williamrandolph committed:
Precompute character run automaton

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
Add a few tests

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

albertzaharovits commented:
@williamrandolph 

> >  do we plan to prevent users from creating restricted indices by mistake?

> The only way for a user to create a system/restricted index is to define it in a custom plugin, so I don't think we need to worry about accidental restricted indices.

Ah, yes, core deals with preventing accidental manipulation of system indices.
I need to brush off my understanding of all of that.

Another tangential question.
This PR makes roles that access system indices require to "tick" the `allow_restricted_indices` role flag. But roles still need to be defined, and index names (or wildcards) be known.

The question: should we automatically be creating roles for these features using system indices? Granted we already have hard-coded roles for some.
I'm trying to avoid having to expose the role flag (including the docs) to users more than it's already exposed.

albertzaharovits commented:
Yet another question: should we emit warnings for user roles in 7.x that cover system indices, and would hence need to be updated to work on 8?

albertzaharovits reviewed (approved):
LGTM

Does what it aims: roles need the `allow_restricted_indices` flag be set in order to authorize users to access system indices. If system indices change over time (across versions or in the same version but by toggling features enabled/disabled), existing user-defined roles (not the ones we hard-code internally) need to be updated.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:
```diff
@@ -281,6 +279,7 @@
                         new RoleDescriptor.IndicesPrivileges[] {
                                 RoleDescriptor.IndicesPrivileges.builder()
                                         .indices(".ml-anomalies*", ".ml-notifications*", ".ml-state*", ".ml-meta*", ".ml-stats-*")
+                                        .allowRestrictedIndices(true)
```
I don't think `machine_learning_admin` should need to access system indices.  The `.ml` indices they are given access to (in the line above) are all hidden indices.  The idea is that ML APIs are used to manipulate the documents in system indices.

If there's a test that fails when `machine_learning_admin` cannot access restricted indices then let me know, because that's probably a bug that needs fixing.

droberts195 replied:
> The idea is that ML APIs are used to manipulate the documents in system indices.

And to complete the picture, when the ML APIs access the system indices they should be doing that as the internal `_xpack` user, which _does_ now need restricted indices allowed.

albertzaharovits replied:
Good catch @droberts195!

Looks like only `.ml-meta*`, `.ml-config*` and `.ml-inference-*` (yes, only the .ml-inference pattern has a dash appended) are registered as system indices, hence the above role change is not needed.
Indeed if this is added to pass a test, it is probably a bug.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:
```diff
@@ -368,6 +370,7 @@
                 .put("enrich_user", new RoleDescriptor("enrich_user", new String[]{ "manage_enrich", "manage_ingest_pipelines", "monitor" },
                         new RoleDescriptor.IndicesPrivileges[]{  RoleDescriptor.IndicesPrivileges.builder()
                             .indices(".enrich-*")
+                            .allowRestrictedIndices(true)
```
I am not sure about this one.  The other externally visible built-in roles that are now being given access to restricted indices are where it's some process owned by Elastic but not Elasticsearch itself that's accessing the system index.

But in the case of enrich I would have though it should be like ML where users call enrich APIs and then the enrich plugin accesses system its system indices to perform the implementation details as the `_xpack` user.  Maybe that's not how it works today, but if so that seems like a code smell where internal information is being unnecessarily exposed.

albertzaharovits replied:
You are right @droberts195 users should not get access to `.enrich-*` indices.
Like in all the other cases (ml, transform, watcher, security, etc) APIs should use special internal "origin" access privilege to access the system index and not require the user have that privilege.

I will raise a bug issue.
In scope of this PR, however, I think this role definition is correct because unfortunately the `.enrich-*` namespace is registered as system.

albertzaharovits replied:
Raised https://github.com/elastic/elasticsearch/pull/76461

williamrandolph committed:
Remove unneeded restricted index access

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
Roll back changes to reserved roles to see what breaks in tests

williamrandolph commented:
@albertzaharovits 
> should we automatically be creating roles for these features using system indices? Granted we already have hard-coded roles for some.

You know, the update here was part of the work that Jay did originally, and I'm not sure what breaks if we take it out. I'm going to back out some changes to the reserved roles to see where tests break and get you a better answer to this question.

> should we emit warnings for user roles in 7.x that cover system indices, and would hence need to be updated to work on 8?

That's a good idea, and I'll make an issue for it as a follow-up.

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph commented:
I spoke to Jay the other day and he didn't have a reason for allowing restricted indices on all of those feature and application roles. We have plenty of test coverage and it all still passes, so I'm going to assume that the "allow restricted indices" change was made by mistake.

I still need to make the issue for a deprecation warning on user roles that try to set access on system indices, but after that I think we should be ready to merge.
- [ ] make issue for deprecating user roles that set privileges on system indices

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
Remove allow restricted indices from client test role

williamrandolph committed:
Try using an admin client for cleaning test state

williamrandolph committed:
checkstyle

williamrandolph committed:
Merge branch 'master' into system_indices_restricted_indices

williamrandolph committed:
Give admin client access to truststore settings

williamrandolph committed:
Add explanatory comments and use warn-level logging for reset failures
